### PR TITLE
Refactor statistics

### LINKED ISSUE #417: Client can only view statistics of planets they have visited
Not sure when this started, it's not new to the latest version and at the very least happens in 0.8.19.7863 with Nebula 0.3.1.

I'm fairly certain it was working in 0.7.x, so maybe it broke in the first 0.8 update and I just didn't notice it.

### PR DESCRIPTION
**Pull request by starfi5h**
Clients now store factory index infomation from the host in StatisticsManager, separte from those in clients' GameData.

- StatisticsManager : store data from the host (factoryCount, planetData <-> factoryIndex relationship)
Access by FactoryCount, GetPlanetData(int factoryIndex), GetFactoryIndex(PlanetData planet)
- UIStatisticsWindow_Transpiler : make update functions refer to StatisticsManager instead

Fix #417, #297

### CODE REVIEW

File: NebulaPatcher/Patches/Dynamic/UIStatisticsWindow_Patch.cs [hunk 1]
```diff
@@ -30,5 +30,13 @@ public static void _OnClose_Postfix()
                 Multiplayer.Session.Network.SendPacket(new StatisticsRequestEvent(StatisticEvent.WindowClosed));
             }
         }
+
+        [HarmonyPrefix]
+        [HarmonyPatch(nameof(UIStatisticsWindow.AddStatGroup))]
+        public static bool AddStatGroup_Prefix(int __0, ProductionStatistics ___productionStat)
+        {
+            //Skip when StatisticsDataPacket hasn't arrived yet
+            return (__0 >= 0 && ___productionStat.factoryStatPool[__0] != null);
```
Comment (PhantomGamers) [inline NebulaPatcher/Patches/Dynamic/UIStatisticsWindow_Patch.cs:39]: Should `__0 < ___productionStat.factoryStatPool.Length` be checked as well?

**Review by PhantomGamers — COMMENTED** [review]
Other than that this seems good to me

File: NebulaPatcher/Patches/Dynamic/UIStatisticsWindow_Patch.cs (hunk 1 above)
Comment (starfi5h) [reply NebulaPatcher/Patches/Dynamic/UIStatisticsWindow_Patch.cs:39]: factoryStatPool is assigned in ``ProductionStatistics.Init(GameData)`` :  
``this.factoryStatPool = new FactoryProductionStat[this.gameData.factories.Length];``
and gameData.factories.Length should be equal or greater than the number of planets, so client and host should have the same factoryStatPool.Length
``__0``, PlanetData.factoryIndex that assigned in ``GameData.GetOrCreateFactory(PlanetData)``, is the index of the PlanetFactory of that planet in ``this.factories`` above 

Therefore ``planet.factoryIndex < ___productionStat.factoryStatPool.Length`` holds  as long as these mechanisms are not changed

**Comment by starfi5h** [comment]
StatisticsManager workflow for future maintainers:

1. Client open the UI statistics panel
-> Send ``StatisticsRequestEvent`` to host, notify window is opened

2. Host ``RegisterPlayer()``, ``ExportAllData()``
-> Send ``StatisticsDataPacket`` to client, which contains factoryCount, planetIds, whole factoryStatPool and techHashedHistory. 

3. Client ``ImportAllData()``
-> Store factoryCount, planetIds in StatisticsManager
-> Replace factoryStatPool with data from host
-> Update the UI statistics panel

4. Host ``CaptureStatisticalSnapshot()``, ``SendBroadcastIfNeeded()``
-> When there is registered player, capture snapshot to record production change in every fix tick
-> broadcast ``StatisticUpdateDataPacket`` to registered players every ``STATISTICS_UPDATE_INTERVAL``
-> If there is a new planetFactory added, send ``StatisticsDataPacket`` instead to update planetIds.

5. Client 
-> Update factoryStatPool with ``StatisticalSnapShot`` received.
-> When UI is closed, send ``StatisticsRequestEvent`` to host, notify window is closed

6. Host ``UnRegisterPlayer()``
-> Unregister player when the window is closed. Stop capture and broadcast when there is no registered player.

**Review by PhantomGamers — APPROVED** [review]

**Review by hubastard — APPROVED** [review]

**Pull request merged**